More and better modals; Social Media sharing

SabreCat opened the pull request:
Moved checklists to Trello: https://trello.com/c/ydzUjKH6/340-social-media-sharing

Feedback requested:
- Current layouts require a lot of `style` attributes to get things to line up correctly, because they use `herobox` and `character-sprites` elements designed to arrange the box used in the site's upper-left corner and in party display. (At least I've figured out how not to do hacky things like negative margins!) Are there better approaches we could use here?

SabreCat committed:
feat(modals): Simplified scroll drop

Conflicts:
	website/views/shared/modals/quests.jade

SabreCat committed:
feat(modals): Small style scroll award

SabreCat committed:
feat(modals): Slim scroll notif
Establishes a new template for modal notifications, to be used across the app including a number of new notifications.

SabreCat committed:
feat(modals): Levelup modal
Introduces a new modal that displays whenever the user levels up.

SabreCat committed:
fix(notifications): Don't double-modal
Corrects behavior of the Notifications controller so that it skips showing the generic levelup modal if it's going to display the quest scroll version.

SabreCat committed:
feat(modals): Hatch Pet modal
Complete except for the planned addition of a heart sprite over the pet.

SabreCat committed:
feat(modals): Grew Mount
A modal now appears when you raise a pet into a mount, which allows the user to immediately equip the new mount.

SabreCat committed:
feat(modals): Won Challenge
Draft of the modal that appears when a user wins a challenge. Needs better logic for when to display.

SabreCat committed:
Merge branch 'develop' into sabrecat/more-modals

SabreCat committed:
feat(sharing): Pretty Tweet button for level up

SabreCat committed:
fix(test): Update push notif string

SabreCat committed:
Merge branch 'develop' into sabrecat/more-modals

SabreCat committed:
feat(sharing): Load widgets from service

SabreCat committed:
fix(sharing): Needed return object

SabreCat committed:
fix(sharing): Better social button positioning

SabreCat committed:
Merge branch 'develop' into sabrecat/more-modals

Conflicts:
	common/locales/en/challenge.json

SabreCat committed:
feat(sharing): Opt Out WIP
First draft of support for opting out of future modal display on specific events, and adds the second barebones target page for sharing.

SabreCat committed:
fix(sharing): Add social service to Inventory
Also fixes accidental 'only' in tests.

SabreCat committed:
feat(sharing): More opt outs
Also fixes a bug where "Display Now" didn't work for pets with a space in their name.

SabreCat committed:
feat(sharing): Buttons for more modals

SabreCat committed:
Merge branch 'develop' into sabrecat/more-modals

SabreCat committed:
WIP(sharing): Incomplete landing page
First stab at creating landing pages for inbound social traffic. May need a different solution.

crookedneighbor commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -53,6 +60,19 @@ habitrpg.controller('NotificationCtrl',
     });
 
+    $rootScope.$watch('user.stats.lvl', function(after, before) {
+      if (after === before) return; 
```
Shouldn't this be `if (before >= after) return;`?

SabreCat replied:
Ah, sure. That'd mean we can remove an extra `if` at line 69, as well. I thought I recalled watches acting weird if you didn't have a strict-equality check at the beginning, but maybe we're safe when we're working with a numeric attribute no user will ever lack.

SabreCat commented on website/public/js/services/socialServices.js:
```diff
@@ -0,0 +1,41 @@
+      }
+
+      // Tumblr
+      $.getScript('https://assets.tumblr.com/share-button.js');
```
It would be lovely if Tumblr has a way, like Twitter and Facebook, to rebuild widgets with the script already loaded. But I couldn't find one. So this results in the Tumblr script potentially getting loaded into memory multiple times over the course of the session. Maybe we can reverse-engineer `share-button.js` a bit? :P

crookedneighbor replied:
Can't you set a variable `tumblrIsLoaded` and check for that before loading?

SabreCat replied:
That'd prevent fetching the script more than once, but it wouldn't activate the piece that goes and sets up the button.

SabreCat commented on website/views/social/level-up.jade:
```diff
@@ -0,0 +1,27 @@
+    title Habitica | Your Life the Role Playing Game
+    script(type='text/javascript').
+      window.env = !{JSON.stringify(env._.pick(env, env.clientVars))};
+    != env.getManifestFiles("tmp_static_front")
```
Adding these three lines is when `social/level-up` started redirecting to `static/front`.

SabreCat replied:
Maybe I need a new set of manifest data instead of trying to reuse `tmp_static_front`?

SabreCat commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -53,6 +60,19 @@ habitrpg.controller('NotificationCtrl',
        Notification.mp(mana);
     });
 
+    $rootScope.$watch('user.stats.lvl', function(after, before) {
```
I tried to write a test for this, but the `$watch` function didn't seem to run upon a `scope.$digest()` during tests.

crookedneighbor committed:
Create a social layout

crookedneighbor committed:
Fix auto-redirect

crookedneighbor committed:
Merge pull request #6147 from crookedneighbor/modals_adjustments

Create a social layout

SabreCat committed:
Merge branch 'develop' into sabrecat/more-modals

SabreCat committed:
WIP(sharing): Draft functional landing page

SabreCat committed:
fix(sharing): Landing page tweaks

SabreCat committed:
fix(sharing): Wording tweaks

SabreCat committed:
feat(sharing): Streak Achievement

SabreCat committed:
Merge branch 'develop' into sabrecat/more-modals

Conflicts:
	common/dist/sprites/spritesmith-largeSprites-0.png
	common/locales/en/front.json

SabreCat committed:
feat(sharing): More achievements

SabreCat committed:
feat(sharing): Remaining cheevo modals

SabreCat committed:
fix(sharing): Use env BASE_URL

SabreCat committed:
fix(sharing): Correct Jade interpolation

SabreCat committed:
fix(sharing): Better Tweet syntax
Angular bindings would fail to parse in Tweets if the expression evaluated to something with a space in it. Adding quotes stretches the Tweet by two characters, but addresses the issue. Commit also takes out a spurious console message from the Notifications service.

SabreCat committed:
fix(sharing): Push ng-binding into Jade

SabreCat committed:
fix(sharing): Translatable share strings

crookedneighbor commented on common/locales/en/front.json:
```diff
@@ -74,7 +74,7 @@
     "invalidEmail": "A valid email address is required in order to perform a password reset.",
     "irishfeet123Quote": "I've had horrible habits with clearing my place completely after meals and leaving cups all over the place. [Habitica] has cured that!",
-    "joinOthers": "Join 850,000 people making it fun to achieve goals!",
+    "joinOthers": "Join 900,000 people making it fun to achieve goals!",
```
This should be string with an interpolation:

```
Join <%= number %> people making it fun to achieve goals!
```

crookedneighbor commented on common/locales/en/front.json:
```diff
@@ -182,5 +182,11 @@
+    "imagine1": "Imagine if improving your life were as fun as playing a game.",
+    "landingCopy1": "Advance in the game by completing your real-life tasks.",
+    "landingCopy2": "Battle monsters with friends to stay accountable to your goals.",
+    "landingCopy3": "Join over 900,000 people having fun as they improve their lives.",
```
Same here. 

```
Join over <%= number %> people having fun as they improve their lives.
```

crookedneighbor commented on test/spec/controllers/inventoryCtrlSpec.js:
```diff
@@ -53,14 +56,36 @@ describe('Inventory Controller', function() {
+      scope.chooseEgg('Cactus');
+      scope.choosePotion('Base');
+
+      expect(rootScope.openModal).to.not.be.called;
```
Excellent tests!

SabreCat replied:
Needs more of them, though. See my comment above about trouble with `$watch` on some of it.

crookedneighbor commented on website/public/js/controllers/inventoryCtrl.js:
```diff
@@ -112,6 +115,10 @@ habitrpg.controller("InventoryCtrl",
       if (!$window.confirm(window.env.t('hatchAPot', {potion: potName, egg: eggName}))) return;
       user.ops.hatch({params:{egg:egg.key, hatchingPotion:potion.key}});
+      if (!user.preferences.suppressModals.hatchPet) {
+        $rootScope.hatchedPet = {egg: eggName, potion: potName, eggKey: egg.key, pet: 'Pet-' + egg.key + '-' + potion.key};
```
Why are we putting this on rootscope? Couldn't you just insert it into the scope of the modal?

SabreCat replied:
How does that go? Use `$scope` instead of `$rootScope`? Though wouldn't that be on `inventoryCtrl`'s scope? (ugh, scopes)

crookedneighbor replied:
Let me see if I can find the thing I'm thinking about.

crookedneighbor replied:
Yeah, you [inject it](http://stackoverflow.com/a/26747669/2601552) using [the resolve option](https://github.com/HabitRPG/habitrpg/blob/develop/website/public/js/controllers/rootCtrl.js#L135)

crookedneighbor replied:
We really abuse `$rootScope` in the app :cry: 

If I remember correctly, objects on `$rootScope` don't get garbage collected as easily as those on `$scope`

SabreCat replied:
Another bad pattern I've absorbed by this being my first major dev project. Add it to the pile. :disappointed:

crookedneighbor replied:
It's not your fault. The app has been in this state for a while now. Bad code begets bad code...

crookedneighbor commented on website/public/js/controllers/inventoryCtrl.js:
```diff
@@ -152,12 +160,16 @@ habitrpg.controller("InventoryCtrl",
 
         _updateDropAnimalCount(user.items);
+        if ($scope.mountCount > startingMounts && !user.preferences.suppressModals.raisePet) {
+          $rootScope.raisedPet = {displayName: petDisplayName, spriteName: pet, egg: egg, potion: potion}
```
Same question here, I think we can just insert this into the scope of the modal.

crookedneighbor commented on website/public/js/controllers/notificationCtrl.js:
```diff
@@ -53,6 +60,23 @@ habitrpg.controller('NotificationCtrl',
+      Notification.lvl();
+      $rootScope.playSound('Level_Up');
+      if (User.user._tmp && User.user._tmp.drop && (User.user._tmp.drop.type === 'Quest')) return;
+      if (after === 3) return; // Drop system unlock. FIXME can we do this without hardcoding?
```
You need a map. IE:

```
var skippableLevels = {
  '3': 'drop system',
  '10':  'class system',
  '50': 'orb of rebirth'
}

if (skippableLevels['' + after]) return;
```

crookedneighbor replied:
the `'' + after` bit is to cast the number as a string.

SabreCat replied:
OK. Right in the same controller, then? I guess that would be easier to update, though it'd still be nice to have the milestone list in a more central place.

crookedneighbor replied:
For now it should be in the controller. If there's a need to have it multiple places, we can put it in a milestones service.

crookedneighbor commented on website/public/js/services/notificationServices.js:
```diff
@@ -108,7 +108,6 @@ angular.module("habitrpg").factory("Notification",
   var stack_topright = {"dir1": "down", "dir2": "left", "spacing1": 15, "spacing2": 15, "firstpos1": 60};
 
   function _notify(html, type, icon) {
-      console.log(type)
```
ABOUT TIME :)

crookedneighbor commented on website/views/shared/modals/achievements.jade:
```diff
@@ -1,99 +1,134 @@
+mixin achievementAvatar(badge,margin)
+  .container-fluid
+    .row
+      .col-xs-4(style='padding:0')
```
I'm not a fan of these inline styles. :/

SabreCat replied:
Neither am I; I commented on that to some extent in the PR description. But so many things needed fiddly little tweaks that it seemed like no matter how I implemented it, it'd add a bunch of ugly clutter somewhere. Suggestions?

crookedneighbor commented on website/views/social/achievement.jade:
```diff
@@ -0,0 +1,9 @@
+block extraHead
+  - var socialPage = 'achievement'
+  - var socialTitle = 'Earned an Achievement!'
+  - var socialDescription = 'I earned a new Achievement on Habitica for working to improve my life!'
```
Did we want to make these translatable?

SabreCat replied:
Ah, can we? I'm fuzzy on how that works. This is something that'd get crawled and displayed when somebody examines a Tweet (or FB share, etc.) on a timeline. Would the meta tags get translated then?

crookedneighbor replied:
I'm also fuzzy on that.

Alys commented on website/public/js/controllers/inventoryCtrl.js:
```diff
@@ -152,12 +160,16 @@ habitrpg.controller("InventoryCtrl",
         $scope.selectedFood = null;
 
         _updateDropAnimalCount(user.items);
+        if ($scope.mountCount > startingMounts && !user.preferences.suppressModals.raisePet) {
```
Using `mountCount` prevents the modal being shown when quest pets are raised to mounts (mountCount is for tracking Mount Master progress). Elsewhere, `statCalc.totalCount(profile.items.mounts)` is used for counting all mounts.
